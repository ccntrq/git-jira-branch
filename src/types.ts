import * as Schema from '@effect/schema/Schema';
import {Brand, Data, Match, type Option, pipe} from 'effect';
import {dual} from 'effect/Function';

export type JiraApiUrl = string & Brand.Brand<'JiraApiUrl'>;
export const JiraApiUrl = Brand.refined<JiraApiUrl>(
  (s: string) => s.startsWith('https://') || s.startsWith('http://'),
  (s: string) => Brand.error(`'${s}' is not a valid Jira API URL`),
);
export type JiraPat = string & Brand.Brand<'JiraPat'>;
export const JiraPat = Brand.refined<JiraPat>(
  (s: string) => s.length > 0,
  () => Brand.error('Provided value is not a valid Jira PAT'),
);
export type JiraApiToken = string & Brand.Brand<'JiraApiToken'>;
export const JiraApiToken = Brand.refined<JiraApiToken>(
  (s: string) => s.length > 0,
  () => Brand.error('Provided value is not a valid Jira API token'),
);
export type JiraUserEmail = string & Brand.Brand<'JiraUserEmail'>;
export const JiraUserEmail = Brand.refined<JiraUserEmail>(
  // NOTE: This validation does not strictly adhere to the email RFC 2822. It is
  // however sufficient to reject inputs that are obviously not valid email
  // addresses while allowing potentially valid ones.
  (s: string) => /^.+@.+\..+$/.test(s),
  (s: string) => Brand.error(`'${s}' is not a valid email address`),
);
export type JiraKeyPrefix = string & Brand.Brand<'JiraKeyPrefix'>;
export const JiraKeyPrefix = Brand.refined<JiraKeyPrefix>(
  (s: string) => s.length > 0,
  (s: string) => Brand.error(`'${s}' is not a valid Jira key prefix`),
);

export type JiraAuth = Data.TaggedEnum<{
  JiraDataCenterAuth: {jiraPat: JiraPat};
  JiraCloudAuth: {jiraUserEmail: JiraUserEmail; jiraApiToken: JiraApiToken};
}>;

export const {JiraDataCenterAuth, JiraCloudAuth} = Data.taggedEnum<JiraAuth>();

export type AppConfig = {
  jiraApiUrl: JiraApiUrl;
  defaultJiraKeyPrefix: Option.Option<JiraKeyPrefix>;
  jiraAuth: JiraAuth;
};

export interface GitBranch {
  name: string;
  isCurrent: boolean;
}

export const GitBranch = Data.case<GitBranch>();

export const JiraIssuetypeSchema = Schema.struct({
  name: Schema.string,
});

export type JiraIssuetype = Schema.Schema.Type<typeof JiraIssuetypeSchema>;

// TODO: transform nullable values into optionals
export const JiraIssueSchema = Schema.struct({
  key: Schema.string,
  fields: Schema.struct({
    summary: Schema.string,
    issuetype: JiraIssuetypeSchema,
    status: Schema.struct({
      name: Schema.string,
    }),
    description: Schema.nullable(Schema.string),
    assignee: Schema.nullable(
      Schema.struct({
        displayName: Schema.string,
      }),
    ),
    creator: Schema.struct({
      displayName: Schema.string,
    }),
  }),
});

export type JiraIssue = Schema.Schema.Type<typeof JiraIssueSchema>;

export type GitCreateJiraBranchResult = Data.TaggedEnum<{
  CreatedBranch: {branch: string};
  SwitchedBranch: {branch: string};
  ResetBranch: {branch: string};
}>;

export const {CreatedBranch, SwitchedBranch, ResetBranch} =
  Data.taggedEnum<GitCreateJiraBranchResult>();

export const matchGitCreateJiraBranchResult: {
  <A>(
    result: GitCreateJiraBranchResult,
    matcher: {
      onCreatedBranch: (branch: string) => A;
      onSwitchedBranch: (branch: string) => A;
      onResetBranch: (branch: string) => A;
    },
  ): A;
  <A>(matcher: {
    onCreatedBranch: (branch: string) => A;
    onSwitchedBranch: (branch: string) => A;
    onResetBranch: (branch: string) => A;
  }): (result: GitCreateJiraBranchResult) => A;
} = dual(
  2,
  <A>(
    result: GitCreateJiraBranchResult,
    matcher: {
      onCreatedBranch: (branch: string) => A;
      onSwitchedBranch: (branch: string) => A;
      onResetBranch: (branch: string) => A;
    },
  ): A =>
    pipe(
      Match.type<GitCreateJiraBranchResult>(),
      Match.tag('CreatedBranch', ({branch}) => matcher.onCreatedBranch(branch)),
      Match.tag('SwitchedBranch', ({branch}) =>
        matcher.onSwitchedBranch(branch),
      ),
      Match.tag('ResetBranch', ({branch}) => matcher.onResetBranch(branch)),
      Match.exhaustive,
    )(result) as A,
);

export type GitJiraBranchError =
  | AppConfigError
  | UsageError
  | GitExecError
  | JiraApiError;

export declare namespace GitJiraBranchError {
  export interface Proto {
    readonly _tag: string;
    readonly message: string;
  }

  export type ProvidedFields = '_tag';
}

const makeError =
  <E extends GitJiraBranchError.Proto>(tag: E['_tag']) =>
  (args: Omit<E, GitJiraBranchError.ProvidedFields>): E =>
    Data.struct({
      _tag: tag,
      ...args,
    } as E);

export interface GitExecError extends GitJiraBranchError.Proto {
  readonly _tag: 'GitExecError';
}

export const GitExecError = makeError<GitExecError>('GitExecError');

export interface UsageError extends GitJiraBranchError.Proto {
  readonly _tag: 'UsageError';
}

export const UsageError = makeError<UsageError>('UsageError');

export interface AppConfigError extends GitJiraBranchError.Proto {
  readonly _tag: 'AppConfigError';
}
export const AppConfigError = makeError<AppConfigError>('AppConfigError');

export const isAppConfigError = (e: unknown): e is AppConfigError =>
  e instanceof Object && '_tag' in e && e._tag === 'AppConfigError';

export interface JiraApiError extends GitJiraBranchError.Proto {
  readonly _tag: 'JiraApiError';
}
export const JiraApiError = makeError<JiraApiError>('JiraApiError');
