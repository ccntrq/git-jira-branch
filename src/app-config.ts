import * as Doc from '@effect/printer/Doc';
import {
  type Brand,
  Config,
  ConfigError,
  Context,
  Effect,
  Either,
  Layer,
  flow,
} from 'effect';
import {InvalidData} from 'effect/ConfigError';

import {
  type AppConfig,
  AppConfigError,
  JiraApiToken,
  JiraApiUrl,
  JiraCloudAuth,
  JiraDataCenterAuth,
  JiraKeyPrefix,
  JiraPat,
  JiraUserEmail,
} from './types';

const mapBrandOrFail =
  <T extends Brand.Brand<string>>(brand: Brand.Brand.Constructor<T>) =>
  (config: Config.Config<Brand.Brand.Unbranded<T>>) =>
    config.pipe(
      Config.mapOrFail((v) =>
        brand.either(v).pipe(Either.mapLeft(brandErrorToConfigError)),
      ),
    );

const brandErrorToConfigError = (
  e: Brand.Brand.BrandErrors,
): ConfigError.ConfigError =>
  InvalidData([], e.map((_) => _.message).join(' '));

const jiraCloudAuthConfig = Config.all({
  jiraApiToken: Config.string('JIRA_API_TOKEN').pipe(
    mapBrandOrFail(JiraApiToken),
  ),
  jiraUserEmail: Config.string('JIRA_USER_EMAIL').pipe(
    mapBrandOrFail(JiraUserEmail),
  ),
}).pipe(Config.map(JiraCloudAuth));

const jiraDataCenterAuthConfig = Config.all({
  jiraPat: Config.string('JIRA_PAT').pipe(mapBrandOrFail(JiraPat)),
}).pipe(Config.map(JiraDataCenterAuth));

const appConfig = Config.all({
  jiraApiUrl: Config.string('JIRA_API_URL').pipe(mapBrandOrFail(JiraApiUrl)),
  defaultJiraKeyPrefix: Config.option(
    Config.string('JIRA_KEY_PREFIX').pipe(mapBrandOrFail(JiraKeyPrefix)),
  ),
  jiraAuth: Config.orElse(jiraCloudAuthConfig, () => jiraDataCenterAuthConfig),
});

const collectConfigErrorMessagesReducer: ConfigError.ConfigErrorReducer<
  undefined,
  Doc.Doc<never>
> = (() => {
  const collectMessage = (
    _: undefined,
    _path: Array<string>,
    message: string,
  ): Doc.Doc<never> => Doc.text(message);

  return {
    andCase: (_, left, right) => Doc.vcat([left, right]),
    orCase: (_, left, right) =>
      Doc.vcat([Doc.indent(left, 2), Doc.text('or'), Doc.indent(right, 2)]),
    invalidDataCase: (_, path, message) =>
      Doc.text(`Expected valid input for ${path.join('.')} (${message})`),
    missingDataCase: collectMessage,
    sourceUnavailableCase: collectMessage,
    unsupportedCase: collectMessage,
  };
})();

// biome-ignore lint/complexity/noStaticOnlyClass: okay for AppConfigService
export class AppConfigService extends Context.Tag('AppConfigService')<
  AppConfigService,
  {
    readonly getAppConfig: Effect.Effect<AppConfig, AppConfigError, never>;
  }
>() {
  public static readonly Live = Layer.succeed(
    AppConfigService,
    AppConfigService.of({
      getAppConfig: appConfig.pipe(
        Effect.mapError(
          flow(
            ConfigError.reduceWithContext(
              void 0,
              collectConfigErrorMessagesReducer,
            ),
            Doc.render({style: 'pretty'}),
            (message) => AppConfigError({message}),
          ),
        ),
      ),
    }),
  );
}
