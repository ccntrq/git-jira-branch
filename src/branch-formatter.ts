import type {GitBranch} from './types';

import {Doc} from '@effect/printer';
import {green} from '@effect/printer-ansi/Ansi';
import {render} from '@effect/printer-ansi/AnsiDoc';
import {Chunk, pipe} from 'effect';

export const formatBranches = (branches: Chunk.Chunk<GitBranch>): string => {
  const hasCurrent = Chunk.some(branches, (_) => _.isCurrent);
  return pipe(
    branches,
    Chunk.map((branch) =>
      branch.isCurrent
        ? pipe(Doc.text('*'), Doc.catWithSpace(mkGreen(Doc.text(branch.name))))
        : Doc.indent(Doc.text(branch.name), hasCurrent ? 2 : 0),
    ),
    Doc.vcat,
    render({style: 'pretty'}),
  );
};

const mkGreen = Doc.annotate(green);
