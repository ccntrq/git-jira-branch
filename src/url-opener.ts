import * as ShellCommand from '@effect/platform/Command';
import type * as CommandExecutor from '@effect/platform/CommandExecutor';
import {Effect, pipe} from 'effect';
import {UsageError} from './types';

export const openUrl = (
  url: string,
): Effect.Effect<void, UsageError, CommandExecutor.CommandExecutor> =>
  Effect.gen(function* () {
    const platform = yield* getPlatform();

    if (platform === 'linux') {
      return yield* xdgOpen(url);
    }

    // TODO: add support for opening urls on other platforms
    return yield* Effect.fail(
      UsageError({
        message: `Opening urls on ${platform} is not yet supported.`,
      }),
    );
  });

const getPlatform = (): Effect.Effect<NodeJS.Platform> =>
  Effect.succeed(process.platform);

const xdgOpen = (
  url: string,
): Effect.Effect<void, UsageError, CommandExecutor.CommandExecutor> =>
  pipe(
    ShellCommand.make('xdg-open', url),
    ShellCommand.exitCode,
    Effect.mapError(() =>
      UsageError({message: 'Browser could not be opened.'}),
    ),
    Effect.map(() => void 0),
  );
