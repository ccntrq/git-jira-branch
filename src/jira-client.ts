import * as Http from '@effect/platform/HttpClient';
import * as ArrayFormatter from '@effect/schema/ArrayFormatter';
import type * as ParseResult from '@effect/schema/ParseResult';
import {Context, Effect, Layer} from 'effect';

import {AppConfigService} from './app-config';
import {
  type AppConfigError,
  JiraApiError,
  type JiraAuth,
  type JiraIssue,
  JiraIssueSchema,
} from './types';

export class JiraClient extends Context.Tag('JiraClient')<
  JiraClient,
  {
    readonly getJiraIssue: (
      issueKey: string,
    ) => Effect.Effect<JiraIssue, AppConfigError | JiraApiError>;
  }
>() {}

export const JiraClientLive = Layer.effect(
  JiraClient,
  Effect.all([AppConfigService, Http.client.Client]).pipe(
    Effect.map(([env, httpClient]) =>
      JiraClient.of({
        getJiraIssue: (issueId: string) =>
          Effect.gen(function* () {
            const endPoint = `/rest/api/latest/issue/${issueId}`;
            const {jiraAuth, jiraApiUrl} = yield* env.getAppConfig;

            return yield* Http.request
              .get(jiraApiUrl + endPoint, {
                headers: {
                  Authorization: buildJiraAuthorizationHeader(jiraAuth),
                  Accept: 'application/json',
                },
              })
              .pipe(
                Http.client.filterStatusOk(httpClient),
                Effect.flatMap(Http.response.schemaBodyJson(JiraIssueSchema)),
                Effect.scoped,
                handleJiraClientErrors(issueId),
              );
          }),
      }),
    ),
  ),
);

const buildJiraAuthorizationHeader = (jiraAuth: JiraAuth): string => {
  switch (jiraAuth._tag) {
    case 'JiraCloudAuth':
      return `Basic ${Buffer.from(
        `${jiraAuth.jiraUserEmail}:${jiraAuth.jiraApiToken}`,
      ).toString('base64')}`;
    case 'JiraDataCenterAuth':
      return `Bearer ${jiraAuth.jiraPat}`;
  }
};

const handleJiraClientErrors =
  (issueId: string) =>
  (
    eff: Effect.Effect<
      JiraIssue,
      AppConfigError | ParseResult.ParseError | Http.error.HttpClientError
    >,
  ): Effect.Effect<JiraIssue, AppConfigError | JiraApiError> =>
    eff.pipe(
      Effect.catchTag('ParseError', (e) =>
        Effect.fail(
          JiraApiError({
            message: [
              'Failed to parse ticket response from Jira:',
              ...ArrayFormatter.formatErrorSync(e).map(
                (issue) => `'${issue.path.join(' ')}': '${issue.message}'`,
              ),
            ].join('\n'),
          }),
        ),
      ),
      Effect.catchTag('RequestError', (e) =>
        Effect.fail(
          JiraApiError({
            message: `Failed to make ticket request to Jira. Reason: ${e.reason}`,
          }),
        ),
      ),
      Effect.catchTag('ResponseError', (e) =>
        Effect.fail(
          JiraApiError({
            message:
              e.response.status === 404
                ? `Jira returned status 404. Make sure the ticket with id ${issueId} exists.`
                : `Jira Ticket request returned failure. Reason: ${e.reason}${
                    typeof e.error === 'string' ? ` (${e.error})` : ''
                  } StatusCode: ${e.response.status}`,
          }),
        ),
      ),
    );
