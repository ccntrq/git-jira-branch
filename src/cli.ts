import {ValidationError} from '@effect/cli';
import * as Args from '@effect/cli/Args';
import type {CliApp} from '@effect/cli/CliApp';
import * as Command from '@effect/cli/Command';
import * as HelpDoc from '@effect/cli/HelpDoc';
import * as Span from '@effect/cli/HelpDoc/Span';
import * as Options from '@effect/cli/Options';
import type * as CommandExecutor from '@effect/platform/CommandExecutor';
import {Console, Effect, Option, pipe} from 'effect';
import {compose} from 'effect/Function';

import * as packageJson from '../package.json';
import type {AppConfigService} from './app-config';
import {formatBranches} from './branch-formatter';
import {
  getAssociatedBranches,
  gitCreateJiraBranch,
  ticketInfo,
  ticketInfoForCurrentBranch,
  ticketUrl,
  ticketUrlForCurrentBranch,
} from './core';
import type {GitClient} from './git-client';
import {formatIssue} from './issue-formatter';
import type {JiraClient} from './jira-client';
import {type GitJiraBranchError, matchGitCreateJiraBranchResult} from './types';
import {openUrl} from './url-opener';

// for version and help
const gitJiraBranch = pipe(Command.make('git-jira-branch', {}));

const createCommand = pipe(
  Command.make(
    'create',
    {
      options: Options.all({
        baseBranch: Options.withDescription(
          Options.optional(Options.withAlias(Options.text('base'), 'b')),
          'Base revision to create the new branch from (a branch name, tag or commit SHA)',
        ),
        reset: Options.withDescription(
          Options.withAlias(Options.boolean('reset'), 'r'),
          'Reset the branch if it already exists',
        ),
      }),
      jiraKey: Args.withDescription(
        Args.text({name: 'jira-key'}),
        'The Jira ticket key to create a branch for (e.g. FOOX-1234)',
      ),
    },
    ({options, jiraKey}) => {
      return Effect.flatMap(
        gitCreateJiraBranch(jiraKey, options.baseBranch, options.reset),
        compose(
          matchGitCreateJiraBranchResult({
            onCreatedBranch: (branch) =>
              `Successfully created branch: '${branch}'`,
            onSwitchedBranch: (branch) =>
              `Switched to already existing branch: '${branch}'`,
            onResetBranch: (branch) => `Reset branch: '${branch}'`,
          }),
          Console.log,
        ),
      );
    },
  ),
  Command.withDescription(
    `
Fetches the given Jira ticket and creates an aproriately named branch for it.
The branch type (bug or feat) is determined by the ticket type. The branch name
is based on the ticket summary.`,
  ),
);

const openCommand = pipe(
  Command.make(
    'open',
    {
      jiraKey: Args.withDescription(
        Args.optional(Args.text({name: 'jira-key'})),
        'The Jira key for the ticket to open (e.g. FOOX-1234)',
      ),
    },
    ({jiraKey}) =>
      pipe(
        jiraKey,
        Option.match({
          onSome: (jiraKey) => ticketUrl(jiraKey),
          onNone: () => ticketUrlForCurrentBranch(),
        }),
        Effect.tap((url) =>
          Console.log(`Opening ticket url '${url}' in your default browser...`),
        ),
        Effect.flatMap(openUrl),
      ),
  ),
  Command.withDescription(
    `
Opens the given Jira ticket in your default browser. If no ticket is given the
jira ticket for the current branch is opened.`,
  ),
);

const infoCommand = pipe(
  Command.make(
    'info',
    {
      jiraKey: Args.withDescription(
        Args.optional(Args.text({name: 'jira-key'})),
        'The Jira key for the ticket to get information for (e.g. FOOX-1234)',
      ),
    },
    ({jiraKey}) =>
      pipe(
        jiraKey,
        Option.match({
          onSome: ticketInfo,
          onNone: () => ticketInfoForCurrentBranch(),
        }),
        Effect.map(formatIssue),
        Effect.flatMap(Console.log),
      ),
  ),
  Command.withDescription(
    `
Displays information for the given Jira ticket on your terminal. If no ticket is
provided, it presents information for the Jira ticket associated with the
current branch.`,
  ),
);

const listCommand = pipe(
  Command.make('list', {}, () =>
    pipe(
      getAssociatedBranches(),
      Effect.map(formatBranches),
      Effect.flatMap(Console.log),
    ),
  ),
  Command.withDescription(
    `
Lists all branches that appear to be associated with a Jira ticket.`,
  ),
);

const mainCommand = gitJiraBranch.pipe(
  Command.withSubcommands([
    createCommand,
    openCommand,
    infoCommand,
    listCommand,
  ]),
);

const cli = {
  name: packageJson.name,
  version: packageJson.version,
  summary: Span.text(packageJson.description),
};

export const cliEffect = (
  args: Array<string>,
): Effect.Effect<
  void,
  GitJiraBranchError | ValidationError.ValidationError,
  | CliApp.Environment
  | GitClient
  | AppConfigService
  | JiraClient
  | CommandExecutor.CommandExecutor
> =>
  Command.run(
    mainCommand,
    cli,
  )(args).pipe(
    Effect.tapError((e) => {
      if (ValidationError.isValidationError(e)) {
        // handled and printed by the cli library already
        return Effect.unit;
      }
      return printDocs(HelpDoc.p(Span.error(e.message)));
    }),
  );

const printDocs = (doc: HelpDoc.HelpDoc): Effect.Effect<void> =>
  Console.log(HelpDoc.toAnsiText(doc));
