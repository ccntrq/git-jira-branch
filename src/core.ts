import {Chunk, Effect, Option} from 'effect';
import {constFalse, constant, pipe} from 'effect/Function';
import {isNone} from 'effect/Option';

import {isNoSuchElementException} from 'effect/Cause';
import {AppConfigService} from './app-config';
import {GitClient} from './git-client';
import {JiraClient} from './jira-client';
import {
  type AppConfigError,
  CreatedBranch,
  type GitBranch,
  type GitCreateJiraBranchResult,
  type GitExecError,
  type GitJiraBranchError,
  type JiraIssue,
  type JiraIssuetype,
  ResetBranch,
  SwitchedBranch,
  UsageError,
} from './types';

export const gitCreateJiraBranch = (
  jiraKey: string,
  baseBranch: Option.Option<string>,
  reset: boolean,
): Effect.Effect<
  GitCreateJiraBranchResult,
  GitJiraBranchError,
  AppConfigService | GitClient | JiraClient
> =>
  Effect.gen(function* () {
    const [gitClient, jiraClient] = yield* Effect.all([GitClient, JiraClient]);

    const fullJiraKey = yield* fullKey(jiraKey);
    const associatedBranch = yield* getAssociatedBranch(fullJiraKey);

    if (!reset && Option.isSome(associatedBranch)) {
      yield* Effect.fail(
        UsageError({
          message: `A branch for ticket '${fullJiraKey}' already exists: ${associatedBranch.value.name}`,
        }),
      );
    }

    const branchName = yield* pipe(
      associatedBranch,
      Option.match({
        onNone: () =>
          pipe(
            jiraClient.getJiraIssue(fullJiraKey),
            Effect.map((issue) => jiraIssueToBranchName(issue)),
          ),
        onSome: (branch) => Effect.succeed(branch.name),
      }),
    );
    const resetBranch = reset && Option.isSome(associatedBranch);
    yield* Option.match(baseBranch, {
      onNone: constant(gitClient.createGitBranch),
      onSome: gitClient.createGitBranchFrom,
    })(branchName, resetBranch);

    return (resetBranch ? ResetBranch : CreatedBranch)({branch: branchName});
  });

export const switchBranch = (
  jiraKey: string,
): Effect.Effect<
  SwitchedBranch,
  GitJiraBranchError,
  AppConfigService | GitClient
> =>
  pipe(
    jiraKey,
    fullKey,
    Effect.flatMap((fullJiraKey) =>
      pipe(
        fullJiraKey,
        getAssociatedBranch,
        Effect.flatten,
        Effect.flatMap((branch) =>
          Effect.flatMap(GitClient, (_) => _.switchBranch(branch.name)).pipe(
            Effect.map((_) => SwitchedBranch({branch: branch.name})),
          ),
        ),
        Effect.catchIf(isNoSuchElementException, () =>
          Effect.fail(
            UsageError({
              message: `No branch associated with Jira ticket '${fullJiraKey}'`,
            }),
          ),
        ),
      ),
    ),
  );

const getJiraKeyFromCurrentBranch = (): Effect.Effect<
  string,
  UsageError | GitExecError,
  GitClient
> =>
  pipe(
    GitClient,
    Effect.flatMap((_) => _.getCurrentBranch()),
    Effect.map(extractJiraKey),
    Effect.flatMap((key) =>
      Option.match({
        onNone: () =>
          Effect.fail(
            UsageError({message: 'No Jira Key found in current branch'}),
          ),
        onSome: (key: string) => Effect.succeed(key),
      })(key),
    ),
  );

export const ticketUrlForCurrentBranch = (): Effect.Effect<
  string,
  GitJiraBranchError,
  AppConfigService | GitClient
> => getJiraKeyFromCurrentBranch().pipe(Effect.flatMap(ticketUrl));

export const ticketUrl = (
  jiraKey: string,
): Effect.Effect<string, AppConfigError, AppConfigService> =>
  pipe(jiraKey, fullKey, Effect.flatMap(buildTicketUrl));

export const ticketInfoForCurrentBranch = () =>
  getJiraKeyFromCurrentBranch().pipe(Effect.flatMap(ticketInfo));

export const ticketInfo = (jiraKey: string) =>
  Effect.gen(function* () {
    const jiraClient = yield* JiraClient;

    const issue = yield* fullKey(jiraKey).pipe(
      Effect.flatMap(jiraClient.getJiraIssue),
    );

    return issue;
  });

export const getAssociatedBranches = (): Effect.Effect<
  Chunk.Chunk<GitBranch>,
  GitExecError,
  GitClient
> =>
  GitClient.pipe(
    Effect.flatMap((_) => _.listBranches()),
    Effect.map((branches) =>
      Chunk.filter(branches, (branch) =>
        Option.isSome(extractJiraKey(branch.name)),
      ),
    ),
  );

const getAssociatedBranch = (fullJiraKey: string) =>
  pipe(
    GitClient,
    Effect.flatMap((_) => _.listBranches()),
    Effect.map(
      Chunk.findFirst((b) =>
        pipe(
          extractJiraKey(b.name),
          Option.map((key) => key === fullJiraKey),
          Option.getOrElse(constFalse),
        ),
      ),
    ),
  );

const fullKey = (
  jiraKey: string,
): Effect.Effect<string, AppConfigError, AppConfigService> =>
  AppConfigService.pipe(
    Effect.flatMap((_) => _.getAppConfig),
    Effect.map(({defaultJiraKeyPrefix}) =>
      jiraKey.match(/^([a-z]+)-(\d+)$/i) || isNone(defaultJiraKeyPrefix)
        ? jiraKey
        : `${defaultJiraKeyPrefix.value}-${jiraKey}`,
    ),
  );

const slugify = (str: string): string =>
  str
    .toLowerCase()
    .replace(/ä/g, 'ae')
    .replace(/ö/g, 'oe')
    .replace(/ü/g, 'ue')
    .replace(/ß/g, 'ss')
    .replace(/[^a-zA-Z0-9]/g, '-')
    .replace(/-+/g, '-')
    .replace(/^-|-$/g, '');

const jiraIssueToBranchName = (issue: JiraIssue): string => {
  const branchtype = jiraIssuetypeBranchtype(issue.fields.issuetype);
  return `${branchtype}/${issue.key}-${slugify(issue.fields.summary)}`;
};

const jiraIssuetypeBranchtype = (issuetype: JiraIssuetype): string => {
  if (issuetype.name.match(/bug/i)) {
    return 'fix';
  }

  return 'feat';
};

const buildTicketUrl = (
  jiraKey: string,
): Effect.Effect<string, AppConfigError, AppConfigService> =>
  AppConfigService.pipe(
    Effect.flatMap((_) => _.getAppConfig),
    Effect.map(({jiraApiUrl}) => `${jiraApiUrl}/browse/${jiraKey}`),
  );

function extractJiraKey(branchName: string): Option.Option<string> {
  const res = branchName.match(/^(?:\w+\/)?([A-Z]+-\d+)/);
  return Option.fromNullable(res?.[1]);
}
