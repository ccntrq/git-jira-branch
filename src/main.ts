#!/usr/bin/env node

import * as NodeCommandExecutor from '@effect/platform-node/NodeCommandExecutor';
import * as NodeFileSystem from '@effect/platform-node/NodeFileSystem';
import {Cause, Console, Effect, Exit, Layer, pipe} from 'effect';

import {HttpClient} from '@effect/platform';
import {NodeContext} from '@effect/platform-node';
import {cliEffect} from './cli';
import {AppConfigService} from './services/app-config';
import {CustomizationsService} from './services/customizations';
import {GitClientLive} from './services/git-client';
import {JiraClientLive} from './services/jira-client';

const commandExecutorLayer = NodeCommandExecutor.layer.pipe(
  Layer.provide(NodeFileSystem.layer),
);

const gitClientLayer = GitClientLive.pipe(Layer.provide(commandExecutorLayer));

const jiraClientLayer = JiraClientLive.pipe(
  Layer.provide(HttpClient.layer),
  Layer.provide(AppConfigService.Live),
);

const mainLive = Layer.mergeAll(
  commandExecutorLayer,
  gitClientLayer,
  AppConfigService.Live,
  CustomizationsService.Live.pipe(Layer.provide(NodeFileSystem.layer)),
  jiraClientLayer,
  NodeContext.layer,
);

const mainEffect = pipe(
  Effect.sync(() => process.argv),
  Effect.flatMap(cliEffect),
  Effect.provide(mainLive),
  Effect.catchAllDefect((defect) => {
    return Console.error(
      Cause.isRuntimeException(defect)
        ? `RuntimeException defect caught: ${defect.message}`
        : `Unknown defect caught: ${defect}.`,
    ).pipe(Effect.andThen(Effect.fail('Defect')));
  }),
);

Effect.runPromiseExit(mainEffect).then((exit) => {
  if (Exit.isFailure(exit)) {
    process.exit(1);
  }
});
