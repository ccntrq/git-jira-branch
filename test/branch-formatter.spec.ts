import {Chunk} from 'effect';
import {describe, expect, it} from 'vitest';
import {formatBranches} from '../src/branch-formatter';
import {GitBranch} from '../src/types';

describe('branch-formatter', () => {
  describe('formatBranches', () => {
    it('should format branches with current', () =>
      expect(
        formatBranches(
          Chunk.fromIterable([
            GitBranch({
              name: 'feat/DUMMYAPP-123-dummy-issue-summary',
              isCurrent: true,
            }),
            GitBranch({
              name: 'feat/DUMMYAPP-124-another-issue-summary',
              isCurrent: false,
            }),
          ]),
        ),
      ).toMatchSnapshot());
  });

  it('should format branches without current', () =>
    expect(
      formatBranches(
        Chunk.fromIterable([
          GitBranch({
            name: 'feat/DUMMYAPP-123-dummy-issue-summary',
            isCurrent: false,
          }),
          GitBranch({
            name: 'feat/DUMMYAPP-124-another-issue-summary',
            isCurrent: false,
          }),
        ]),
      ),
    ).toMatchSnapshot());
});
