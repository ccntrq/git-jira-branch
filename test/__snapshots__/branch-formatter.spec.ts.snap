// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`branch-formatter > formatBranches > should format branches with current 1`] = `
"* [0;32mfeat/DUMMYAPP-123-dummy-issue-summary[0m
  feat/DUMMYAPP-124-another-issue-summary"
`;

exports[`branch-formatter > should format branches without current 1`] = `
"feat/DUMMYAPP-123-dummy-issue-summary
feat/DUMMYAPP-124-another-issue-summary"
`;
